src/enclave/enclave_signal.c

```diff
@@ -151,33 +183,28 @@ static uint64_t sgxlkl_enclave_signal_handler(
         info.si_addr = (void*)exception_record->address;
         info.si_signo = trap_info.signo;
 
+        /**
+         * The trap is is passed to LKL. If it can be handled, excecution will continue,
```

Comment:
In the abort case, would it print a backtrace? If not, then this PR only partially solves #321, right?

Reply:
Ah, I now understand what you meant in #321. No, you won't get a backtrace in that case. The problem is that oe_backtrace() cannot resolve application symbols.

(If we figured out how to get application symbols, we could always print a backtrace when a signal is received and `SGXLKL_TRACE_SIGNAL=1`. This would be easier than figuring out when LKL is aborting a process due to a signal.)

Reply:
@letmaik I had a quick look at how `oe_backtrace()` works: it uses about 10 lines of code (not `libunwind`) to walk the stack inside the enclave (with some access checks), and then does an ocall to resolve addresses to symbols. The ocall resolves the symbols based on the `libsgxlkl.so` ELF image that was loaded into the enclave.

This isn't sufficient to resolve application symbols, which are loaded dynamically from ELF images inside the enclave. Another complication is that application symbols would have to be resolved in userspace, whereas the mechanism in this PR is implemented in kernel space. In general, I would say that producing stack traces for applications is an orthogonal problem to what SGX-LKL does. It should be possible to do this entirely within the enclave in userspace, without help from SGX-LKL.

You may want to update #321.

Reply:
I updated #321 to be kernel-only. Just to clarify though, if an exception in userspace within the user application is raised and unhandled by the application, will that surface as an unhandled kernel-level exception (with unresolved symbols) as in this PR? A similar question, since we're splitting SGX-LKL into kernel code and userspace code at the moment, what will happen to exceptions in the userspace code that are from SGX-LKL and not from the application? I can't remember exactly what the split is, but I think mounting disks would move to userspace, right?

Reply:
No, because if the signal is not handled by the enclave application, the SGX-LKL exception handler will not see this but only LKL. LKL will then abort the thread/process but this happens entirely within LKL code. The abort will result in a graceful SGX-LKL shutdown that returns an error code.

You will get valid stack traces for userspace code as long as it is part of `libsgxlkl.so`. There was some discussion about splitting userspace into a separate file, and things may change then. It all depends if `oe_backtrace` still knows how to find the debug symbols.